akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/AbstractLeastShardAllocationStrategy.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.internal
+
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.actor.Address
+import akka.annotation.InternalApi
+import akka.cluster.Cluster
+import akka.cluster.Member
+import akka.cluster.MemberStatus
+import akka.cluster.sharding.ShardCoordinator.ActorSystemDependentAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+
+import scala.collection.immutable
+import scala.collection.immutable.SortedSet
+
+/**
+ * Common logic for the least shard allocation strategy implementations
+ *
+ * INTERNAL API
+ */
+private[akka] object AbstractLeastShardAllocationStrategy {
+  import MemberStatus._
+  private val DoNotRebalanceTo: Set[MemberStatus] = Set(Leaving, Exiting, Down)
+  // defer rebalance when nodes are soon becoming up
+  private val AvoidRebalanceWhen: Set[MemberStatus] = Set(Joining, WeaklyUp)
+
+  type AllocationMap = Map[ActorRef, immutable.IndexedSeq[ShardId]]
+
+  implicit object ShardSuitabilityOrdering extends Ordering[(ActorRef, Member, immutable.IndexedSeq[ShardId])] {
+    override def compare(
+        x: (ActorRef, Member, immutable.IndexedSeq[ShardId]),
+        y: (ActorRef, Member, immutable.IndexedSeq[ShardId])): Int = {
+      val (_, memberX, allocatedShardsX) = x
+      val (_, memberY, allocatedShardsY) = y
+      if (memberX.status != memberY.status) {
+        // allocating to nodes that are on their way out of the cluster should always be
+        // the least option
+        // FIXME: more clever logic around states here (prefer up over joining)?
+        DoNotRebalanceTo(memberX.status).compareTo(DoNotRebalanceTo(memberY.status))
```

Comment:
I find this hard to reason/read. 

First, the usage of `apply` to check if it `contains` is, forces the reader to go look what is actually `DoNotRebalanceTo`. We need the extra knowledge that `Set.apply` is the same as `Set.contains`. 

Then using boolean with compareTo. Not so intuitive to me.

Reply:
I guess I can do `<` and return an int, just felt convenient to use the existing boolean compareTo rather than re-implement it.

Reply:
I left it compareTo but hopefully made it easier to read.